consensus_encoding/src/encode/encoders.rs

```diff
@@ -0,0 +1,150 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! Collection of "standard encoders".
+
+/// An encoder for a single byte slice.
+use super::Encoder;
+
+/// An encoder for a single byte slice.
+pub struct BytesEncoder<'sl> {
+    sl: &'sl [u8],
+    done: bool,
+}
+
+impl<'sl> BytesEncoder<'sl> {
+    /// Constructs a byte encoder which encodes the given byte slice, with no length
+    /// prefix.
+    pub fn without_length_prefix(sl: &'sl [u8]) -> Self { Self { sl, done: false } }
+}
+
+impl<'sl> Encoder for BytesEncoder<'sl> {
+    fn advance(&mut self) -> Option<&[u8]> {
+        if self.done {
+            None
+        } else {
+            self.done = true;
+            Some(self.sl)
+        }
+    }
+
+    fn unadvance(&mut self) { self.done = false; }
+}
+
+/// An encoder for a single array.
+pub struct ArrayEncoder<const N: usize> {
+    arr: [u8; N],
+    done: bool,
+}
+
+impl<const N: usize> ArrayEncoder<N> {
+    /// Constructs an encoder which encodes the array with no length prefix.
+    pub fn without_length_prefix(arr: [u8; N]) -> Self { Self { arr, done: false } }
+}
+
+impl<const N: usize> Encoder for ArrayEncoder<N> {
+    fn advance(&mut self) -> Option<&[u8]> {
+        if self.done {
+            None
+        } else {
+            self.done = true;
+            Some(&self.arr)
+        }
+    }
+
+    fn unadvance(&mut self) { self.done = false; }
+}
+
+/// An encoder which encodes two objects, one after the other.
+pub struct Encoder2<A, B> {
+    enc_idx: usize,
+    enc_1: A,
+    enc_2: B,
+}
+
+impl<A, B> Encoder2<A, B> {
+    /// Constructs a new composite encoder.
+    pub fn new(enc_1: A, enc_2: B) -> Self { Self { enc_idx: 0, enc_1, enc_2 } }
+}
+
+impl<A: Encoder, B: Encoder> Encoder for Encoder2<A, B> {
+    fn advance(&mut self) -> Option<&[u8]> {
+        if self.enc_idx == 0 {
+            if let Some(res) = self.enc_1.advance() {
+                return Some(res);
+            }
+            self.enc_idx += 1;
+        }
+        self.enc_2.advance()
+    }
```

Comment:
I'm a bit suss on this logic but for concept ACK its fine. I'll convince myself all paths through this funtion are correct in a later review.

Reply:
Convinced myself you logic is correct, here and in `advance` - nice!